object_oriented_programming.ipynb

```diff
@@ -0,0 +1,1534 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "38c920c6-d21c-41f9-b45c-65352822c3e5",
+   "metadata": {},
+   "source": [
+    "Here's our final class, without any `print` statement – remember, we should try to avoid side-effects when they are **not** necessary:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b4982727-ac70-4b25-9a92-6171228637a8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Rectangle:\n",
+    "    def __init__(self, width, height):\n",
+    "        self.width = width\n",
+    "        self.height = height\n",
+    "        \n",
+    "    def area(self):\n",
+    "        return self.width * self.height\n",
+    "    \n",
+    "    def perimeter(self):\n",
+    "        return 2 * (self.width + self.height)\n",
+    "    \n",
+    "    def __str__(self):\n",
+    "        return f'Rectangle (width={self.width}, height={self.height})'\n",
+    "    \n",
+    "    def __repr__(self):\n",
+    "        return f'Rectangle (width={self.width}, height={self.height})'\n",
+    "    \n",
+    "    def __eq__(self, other):\n",
+    "        if isinstance(other, Rectangle):\n",
+    "            return (self.width, self.height) == (other.width, other.height)\n",
+    "        else:\n",
+    "            return False"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ea2c70e3-e22b-4064-9159-61275d65b143",
+   "metadata": {},
+   "source": [
+    "What about `<`, `>`, `<=`, etc.?\n",
+    "\n",
+    "Again, Python has special methods we can use to provide that functionality.\n",
+    "\n",
+    "These are `__lt__`, `__gt__`, `__le__` methods. There are [many more](https://docs.python.org/3/reference/datamodel.html)!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "64d77b6d-7051-4f4c-a2eb-cece6c6eb58b",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "class Rectangle:\n",
+    "    def __init__(self, width, height):\n",
+    "        self.width = width\n",
+    "        self.height = height\n",
+    "        \n",
+    "    def area(self):\n",
+    "        return self.width * self.height\n",
+    "    \n",
+    "    def perimeter(self):\n",
+    "        return 2 * (self.width + self.height)\n",
+    "    \n",
+    "    def __str__(self):\n",
+    "        return f'Rectangle (width={self.width}, height={self.height})'\n",
+    "    \n",
+    "    def __repr__(self):\n",
+    "        return f'Rectangle (width={self.width}, height={self.height})'\n",
+    "    \n",
+    "    def __eq__(self, other):\n",
+    "        if isinstance(other, Rectangle):\n",
+    "            return (self.width, self.height) == (other.width, other.height)\n",
+    "        else:\n",
+    "            return False\n",
+    "    \n",
+    "    def __lt__(self, other):\n",
```

Comment:
Maybe briefly mention that this is a bit "risky" as rectangles themselves are not ordered in a strictly mathematical sense.